rfcs/address_space_overrides.md

```diff
@@ -0,0 +1,111 @@
+
+## Details
+
+### Address allocation
+
+For the purposes of testing, declare that the following subnets are not `local`
+but instead:
+
+* 127.1.0.0/16: `private`
+* 127.2.0.0/16: `public`
+
+### Browser changes
+
+A new command-line flag (or a collection thereof) is added to each browser (it
+need not be uniform, though that would certainly help with implementation) that
+allows overriding the address space derived from specific IP addresses. For
+example:
+
+```sh
+--test-override-address-space=127.1.0.0/16:private,127.2.0.0/16:public
+```
+
+Test-only code wires this command-line flag to the code in the browser under
+test that determines the address space of a network endpoint.
+
+### Test runner changes
+
+The web platform test runner sets command-line flags on browsers such that the
+browsers implement the custom address space mapping described above during
+tests.
+
+Corresponding entries are added to the hosts file, as generated by
+[make_hosts_file()](https://github.com/web-platform-tests/wpt/blob/master/tools/serve/serve.py#L513-L530):
+
+```
+127.1.0.1 private.web-platform.test
+127.2.0.1 public.web-platform.test
+```
+
+This allows web platform tests to exercise browser behavior in the face of
+`private` and `public` IP addresses, simply by targeting the above domains.
+
+## Risks
+
+If the tests run on an IPv6-only system, the system might not know how to route
+requests to the IPv4 loopback address range? This seems like an extremely remote
+possibility, in which case many other assumptions made by the test framework
+might have to be revisited anyway.
+
+## Alternatives considered
+
+### Use individual IP addresses
+
+Same as above, but allow overriding address spaces per IP address instead of
+per subnet.
+
+The marginal difficulty of overriding per subnet seems low, and provides plenty
+of test IP addresses to exercise intra-address-space, cross-ip-address requests,
+which might come in handy [in the
+future](https://github.com/WICG/cors-rfc1918/pull/1#issuecomment-721110250).
+
+### Use real IP addresses
+
+If we need to test user agent behavior in the face of a variety of IP addresses,
+then... just do it!
+
+This approach would have the web platform test runner configure additional
+loopback network devices and assign them `private` and `public` IP addresses.
+Then we could configure the hosts file to resolve specific domains to those IP
+addresses, and test the whole shebang end to end.
+
+This approach's main pro is that it truly is an end to end test of the feature,
+exercising production code to its full extent. By virtue of requiring no
+test-only hooks, it also requires no work on the part of browser developers.
+
+Its main con is that it requires the test runner to have system privileges.
+Additionally, it requires work to support for each test execution environment.
+Finally, it arguably requires the purchase of a public IP address for our use,
```

Comment:
Does it really require purchasing a public IP address? I am not sure why. Would it be so unacceptable to choose some arbitrary public IP address and redirect it, only for the purpose of the test execution, to localhost? Executing WP tests does not need access to public IP addresses, right? So I don't see room for conflicts.

Reply:
Although the tests themselves don't need to access internet, the runner (especially in setup/teardown phases, e.g. to download pip packages, browser binaries, test prefs, etc.) usually does.

I guess we could carefully set up the redirection after all other setup finishes and right before tests start, and remove it at the end, but that'd require giving root to the test runner; see my reply to the other comment.

Reply:
I'll add some detail to this section to make this point.